docs/content/preview/api/ysql/user-defined-subprograms-and-anon-blocks/language-plpgsql-subprograms/plpgsql-syntax-and-semantics/executable-section/basic-statements/doing-sql-from-plpgsql.md

```diff
@@ -0,0 +1,411 @@
+
+Briefly, the risk occurs when you want to construct the text of a SQL statement programmatically and you want it to contain identifiers and values that become known only at run-time. Naïve programmers simply assemble the text using the ordinary _text_ concatenation operator—but you must _never_ do this. The examples that Internet search finds show you what can go wrong if you do this.
+
+(You can reduce the size of the hit-list to about _300 thousand_ by searching for _PostgreSQL "format()" "SQL injection"_. )
+
+You must _always_ must the _format()_ built-in SQL function when you need to define the text of a dynamic SQL statement at run-time. The example above shows you how to do this.
+{{< /tip >}}
+
+### The "perform" statement
+
+_[See the syntax rule above](./#plpgsql-perform-stmt)_
+
+The _perform_ statement lets you execute a select statement that, as a top-level SQL statement, would produce one or many rows. But, critically, it lets you avoid using the _select into_ statement for one row or the _[query for loop]((../../compound-statements/loop-exit-continue/query-for-loop/))_ (or the _[infinite loop]((../../compound-statements/loop-exit-continue/query-for-loop/))_) for many rows. It's the moral equivalent of this:
+
+```output
+select ... into null;
+```
+
+analogous, in Unix, to sending output to _/Dev/Null_. Try this top-level SQL statement:
+
+```plpgsql
+with s(val) as (select g.val from generate_series(1, 11) as g(val))
+select s.val as k, s.val*10 as v from s;
+```
+
+This, of course, is the result:
+
+```output
+ k | v  
+---+----
+ 1 | 10
+ 2 | 20
+ ......
+ 8 | 80
+ 9 | 90
+```
+
+To use this as the argument for _perform_, you must encapsulate as a subquery in an outer _select_ statement, thus:
+
+```plpgsql\
+select k, v from (
+    with s(val) as (select g.val from generate_series(1, 11) as g(val))
+    select s.val as k, s.val*10 as v from s
+  ) as subqry;
+```
+
+You can now use this in a PL/pgSQL block statement simply by replacing the starting _select_ with _perform_. Do this:
+
+```plpgsql
+do $body$
+begin
+  perform k, v from (
+      with s(val) as (select g.val from generate_series(1, 11) as g(val))
+      select s.val as k, s.val*10 as v from s
+    ) as subqry;
+end;
+$body$;
+```
+
+It finishes silently without error.
+
+This _do_ statement immediately above shows the general case, where the _select_ statement whose results you want to throw away must use a _with clause_.
+
+ This need arises, for example, when you perform multi-table _insert_, _update_ or _delete_ by encapsulating these statements each with a pro-forma _returning_ clause, and in its own _common table expression (hereinafter _CTE_). Then you cause them to execute by selecting the union of these pro-forma results  from the various _CTEs_. But the results are of no interest. They were simply a necessary device.
```

Comment:
missing closing underscore - _common table expression_ (hereinafter _CTE_)
Though CTE was already defined earlier, so maybe no need to redefine here

Reply:
Yes, I already had this earlier in the page:

> The clue is the _[common_table_expression](../../../../../../syntax_resources/grammar_diagrams/#common-table-expression)_ hereinafter _CTE_. 

So I simplified the wording that you called out in this comment thus:

> ...by encapsulating these statements each with a pro-forma _returning_ clause, and in its own _[CTE](../../../../../../syntax_resources/grammar_diagrams/#common-table-expression)_.

The link is the same as in the fragment I just quoted above.